[AArch64][PAC][MC][ELF] Support PAuth ABI compatibility tag

kovdan01 opened the pull request:
Depends on #87545

Emit `GNU_PROPERTY_AARCH64_FEATURE_PAUTH` property in `.note.gnu.property` section depending on
`aarch64-elf-pauthabi-platform` and `aarch64-elf-pauthabi-version` llvm module flags.

kovdan01 committed:
[AArch64][PAC][MC][ELF] Support PAuth ABI compatibility tag

Emit `GNU_PROPERTY_AARCH64_FEATURE_PAUTH` property in
`.note.gnu.property` section depending on
`aarch64-elf-pauthabi-platform` and `aarch64-elf-pauthabi-version` llvm
module flags.

kovdan01 commented:
Buildkite job expectedly ails since the PR depends on #85235 unmerged yet

DanielKristofKiss commented on llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:
```diff
@@ -268,13 +268,24 @@ void AArch64AsmPrinter::emitStartOfAsmFile(Module &M) {
+          M.getModuleFlag("aarch64-elf-pauthabi-version")))
+    PAuthABIVersion = PAV->getZExtValue();
+
+  if ((PAuthABIPlatform == uint64_t(-1)) != (PAuthABIVersion == uint64_t(-1)))
```
could this check be added to the verifier?

kovdan01 replied:
Thanks for suggestion, see ab89fdcb49f9ae6dd2d20cac81cca5204eda51a6

kovdan01 committed:
Merge branch 'main' into pauth-gnuprop-mc

kovdan01 committed:
Rename PAuth ABI tag to PAuth ABI core info (see #85231)

kovdan01 committed:
Move check of module flags to verifier

DanielKristofKiss commented on llvm/lib/IR/Verifier.cpp:
```diff
@@ -1734,8 +1734,29 @@ void Verifier::visitModuleFlags() {
+  }
+
+  if ((PAuthABIPlatform == uint64_t(-1)) != (PAuthABIVersion == uint64_t(-1)))
+    report_fatal_error(
```
```suggestion
    CheckFailed(
```

kovdan01 replied:
Fixed, thanks, see 2f6dc0adbee9e09176e86dd9960b8d353c31decf

DanielKristofKiss reviewed (approved):
Could you consider adding a test for the verifier?

kovdan01 committed:
Address review comments

- Use `CheckFailed` instead of `report_fatal_error` in verifier
- Add tests for verifier

kovdan01 commented:
> Could you consider adding a test for the verifier?

@DanielKristofKiss Thanks for suggestion, see llvm/test/Verifier/module-flags-note-gnu-property-elf-pauthabi.ll added in 2f6dc0adbee9e09176e86dd9960b8d353c31decf

kovdan01 committed:
Fix code formatting

kovdan01 committed:
Merge branch 'main' into pauth-gnuprop-mc
